### PR TITLE
Feature/topiary

### PR DESCRIPTION
**Pull request by snehring**
A pipeline

### CODE REVIEW

File: var/spack/repos/builtin/packages/generax/package.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    homepage = "https://github.com/BenoitMorel/GeneRax"
+    git = "https://github.com/BenoitMorel/GeneRax.git"
+
+    maintainers("snehring")
+
+    version("master", branch="master", submodules=True)
+    version("dev", branch="dev", submodules=True)
+    version("2.0.4", commit="e4fab40f407bdd3b588d3d69a449f8c1be56f9fa", submodules=True)
+    version("2.0.1", commit="edff1caca0fe3f691c386e4b9a990f391dbdcc60", submodules=True)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/generax/package.py]: Are you sure this is the commit you want as it appears to be associated with `2.0.2` based on the diff (https://github.com/BenoitMorel/GeneRax/commit/edff1caca0fe3f691c386e4b9a990f391dbdcc60)?

This is the corresponding commit that updates that line to `2.0.1`:  https://github.com/BenoitMorel/GeneRax/commit/7d16df4c0125c24df9247a7be68ac51f19d68fee.  While this is the commit associated with the `2.0.1` release: https://github.com/BenoitMorel/GeneRax/commit/0623dae55dd602a60faae63e9991fa8d41782456.

**Review by tldahlgren — COMMENTED** [review]
Confirmed the version branches, sha256, and commits as well as home pages.

**Comment by snehring** [comment]
I did a lot of squashing in this mess after debugging some things. It's extremely likely I messed that up. I'll fix it up tomorrow.

File: var/spack/repos/builtin/packages/py-ete3/package.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    maintainers("snehring")
+
+    version("3.1.2", sha256="4fc987b8c529889d6608fab1101f1455cb5cbd42722788de6aea9c7d0a8e59e9")
+
+    depends_on("py-setuptools", type="build")
+
+    depends_on("py-numpy", type=("build", "run"))
+    depends_on("py-pyqt5", type=("build", "run"))
+    depends_on("py-lxml", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-ete3/package.py:27]: These 3 are optional, not sure if we want to make variants for these or just always require them

File: var/spack/repos/builtin/packages/py-itolapi/package.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    homepage = "https://github.com/albertyw/itolapi"
+    pypi = "itolapi/itolapi-4.1.2.tar.gz"
+
+    maintainers("snehring")
+
+    version("4.1.2", sha256="37a866a117a80d3d72a6eb6b2cba30444751c644cc6bc4242f050750375a8397")
+
+    depends_on("py-setuptools", type="build")
+
+    depends_on("py-requests@2.0:3.0", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-itolapi/package.py]:
```suggestion
    depends_on("py-requests@2", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-opentree/package.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    """Library for interacting with Open Tree of Life resources."""
+
+    homepage = "https://github.com/OpenTreeOfLife/python-opentree"
+    pypi = "opentree/opentree-1.0.1.tar.gz"
+
+    maintainers("snehring")
+
+    version("1.0.1", sha256="a765ae37cd72c232c292506c170656aaaa4be81e6de6ccd4845eec01bfe45e4d")
+
+    depends_on("py-setuptools", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-opentree/package.py]:
```suggestion
    depends_on("py-setuptools", type=("build", "run"))
```
Technically it's not used at run-time, but the `setup.py` lists it under `install_requires`, so `pip check` will complain if setuptools isn't found. Might be worth opening a PR to move it to `setup_requires`.

File: var/spack/repos/builtin/packages/py-pastml/package.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    phylogenetic trees.
+    """
+
+    homepage = "https://github.com/evolbioinfo/pastml"
+    pypi = "pastml/pastml-1.9.40.tar.gz"
+
+    maintainers("snehring")
+
+    version("1.9.40", sha256="5334bc8de70a968117240b90d90878ac935be18de6e6e485fb1a8f90cd539fea")
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pastml/package.py:20]: Python 3.8+

File: var/spack/repos/builtin/packages/py-topiary/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-topiary/package.py:1]: Let's name this `py-topiary-asr`. There's already a different topiary package on PyPI. Also see https://github.com/harmslab/topiary/blob/86913d42417bf45ca262e7740672c61a3ed85455/setup.py#L41

File: var/spack/repos/builtin/packages/py-topiary/package.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    maintainers("snehring")
+
+    version("main", branch="main")
+    version("0.9.9", sha256="5601fba92e7add33a3732482426b2c7ef46b0fccc4a4ea11357537e1b937903c")
+
+    depends_on("python@3.8:", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
+
+    depends_on("py-numpy", type="run")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-topiary/package.py]:
```suggestion
    depends_on("py-numpy@:1.21", type="run")
```

File: var/spack/repos/builtin/packages/py-topiary/package.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    depends_on("py-pandas", type="run")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-topiary/package.py]: Missing xlrd and openpyxl from requirements.txt

File: var/spack/repos/builtin/packages/py-topiary/package.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    depends_on("py-matplotlib", type="run")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-topiary/package.py]: Missing biopython

File: var/spack/repos/builtin/packages/py-topiary/package.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    depends_on("py-ete3", type="run")
+    depends_on("py-toytree", type="run")
+    depends_on("py-dendropy", type="run")
+    depends_on("py-pastml", type="run")
+    depends_on("py-opentree", type="run")
+    depends_on("py-tqdm", type="run")
+    depends_on("py-mpi4py", type="run")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-topiary/package.py]: Where does mpi4py come from? Also, these should all be `type=("build", "run")`

File: var/spack/repos/builtin/packages/py-topiary/package.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+    depends_on("blast-plus", type="run")
+    depends_on("muscle5@5.0:", type="run")
+    depends_on("generax@2.0:+mpi", type="run")
+    depends_on("raxml-ng@1.1:", type="run")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-topiary/package.py]: Might be worth adding a comment to explain where these came from

File: var/spack/repos/builtin/packages/py-topiary/package.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    depends_on("openmpi+legacylaunchers", type="run", when="^openmpi schedulers=slurm")
+
+    conflicts("mpich")
+
+    def patch(self):
+        if self.spec.satisfies("^raxml-ng+mpi"):
+            filter_file(
+                r"RAXML_BINARY\s*=\s*\"raxml-ng\"$",
+                'RAXML_BINARY = "raxml-ng-mpi"',
+                "topiary/raxml/_raxml.py",
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-topiary/package.py]:
```suggestion
                join_path("topiary", "raxml", "_raxml.py"),
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-ete3/package.py (hunk 2 above)
Comment (snehring) [reply var/spack/repos/builtin/packages/py-ete3/package.py:27]: I think I was running into build errors without including these.

File: var/spack/repos/builtin/packages/py-topiary/package.py (hunk 9 above)
Comment (snehring) [reply var/spack/repos/builtin/packages/py-topiary/package.py]: The mpi4py dep isn't documented. There's a couple places in the mpi parts of this that import it iirc, at the very least in https://github.com/harmslab/topiary/blob/main/topiary/_private/mpi/_get_hosts.py

File: var/spack/repos/builtin/packages/py-topiary-asr/package.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    maintainers("snehring")
+
+    version("main", branch="main")
+    version("0.9.9", sha256="5601fba92e7add33a3732482426b2c7ef46b0fccc4a4ea11357537e1b937903c")
+
+    depends_on("python@3.8:", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
+
+    depends_on("py-biopython", type=("build", "run"))
+    depends_on("py-numpy", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-topiary-asr/package.py]:
```suggestion
    depends_on("py-numpy@:1.21", type=("build", "run"))
```

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**